docs/source/howto/codes.rst

```diff
@@ -12,6 +12,34 @@ Interfacing external codes
 
 `#3986`_
 
+In order to work with an external simulation code in AiiDA, you need a calculation input plugin (subclassing the :py:class:`~aiida.engine.CalcJob` class) and an output parser plugin (subclassing the :py:class:`~aiida.parsers.Parser` class):
+
+Before starting to write a plugin, check on the `aiida plugin registry <https://aiidateam.github.io/aiida-registry/>`_ whether a plugin for your code is already available.
+
+Design guidelines
+------------------
+
+ * | **Start simple.**
+   | Make use of existing classes like :py:class:`~aiida.orm.nodes.data.dict.Dict`, :py:class:`~aiida.orm.nodes.data.singlefile.SinglefileData`, ...
+   | Write only what is necessary to pass information from and to AiiDA.
+ * | **Don't break data provenance.**
+   | Store *at least* what is needed for full reproducibility.
+ * | **Parse what you want to query for.**
+   | Make a list of which information to:
+
+     #. parse into the database for querying (:py:class:`~aiida.orm.nodes.data.dict.Dict`, ...)
+     #. store in files for safe-keeping (:py:class:`~aiida.orm.nodes.data.singlefile.SinglefileData`, ...)
+     #. leave on the remote computer (:py:class:`~aiida.orm.nodes.data.remote.RemoteData`, ...)
+
+ * | **Expose the full functionality.**
+   | Standardization is good but don't artificially limit the power of a code you are wrapping - or your users will get frustrated.
+   | If the code can do it, there should be *some* way to do it with your plugin.
```

Comment:
> At the risk of devolving in some kind of philosophical debate, I believe this is impractical. This requires a full replication of the API of a code which is an extremely evolved and basically unattainable goal.

I'm afraid I have to disagree again. We are talking about wrapping command line executables here.
It is very straightforward to write an AiiDA plugin, that simply takes an input file (or has a way to accept multiple input files) and provide them to the code.
While this makes your inputs non-queryable (and will provide no automatic validation), it's a way to allow your users to get their work done via AiiDA even if you have not yet created a specific API for writing all aspects of an input file.
Leaving a "backdoor" like this can be very useful and avoid frustration of having to deal with a "crippled" version of the code you are used to work with.

> What if I am only interested in a specific function of a code for a specific purpose? From a problem-oriented perspective: Do I really have to wrap the complete API of a code in order to be able to use it with AiiDA?
> Final remark: We are focusing on solving user problems here. I don't think that users should need to worry about other who users who might get frustrated with their plugin if what they care about is "interfacing external codes" for their own use cases.

Now, with this I agree - as I commented above (and as was probably not clear), I have moved these guidelines from "plugin design" to this section, since this section deals with the writing of CalcJobs.
For a super-short CalcJob howto one could even do without any guidelines on design.
I'm happy to move them to a different home, but I do feel they are important, since they matter when you want to write plugins used by other people.

Reply:
Anyhow, this can be taken care of in the PR that will deal with #3986

Reply:
I agree that the slightly more demanding plugin guidelines should be very clearly separated from the more general and immediate problem of "how to interface code". So as long as we make that distinction with #3986 , I think this is fine.

I get the impression that I may still be missing the point and this is actually related to my previous comment. I agree that *ideally* a plugin should not limit users in their ability to use all capabilities of a code, but I will insist that a complete replication of an API is very hard to achieve.

This discussion is not a blocker for this PR.

Reply:
> a complete replication of an API is very hard to achieve.

That is not my experience. I certainly agree that a *user-friendly, queryable, validated, ...* replication of an entire API is hard to achieve. What I am suggesting here is to have a nice standardized way of doing *some* of what your wrapped code can do, and then leave a backdoor for people to do what they need to do, if it is not covered by your fancy interface (e.g. by passing through input files directly).
Once you notice users using the backdoor often, you can then move the functionality you are using to the "nice" interface.
In my experience, this is rather straightforward, and a better strategy than to block users from using functionality they know that their code has.

Anyhow, perhaps this requires one or two sentences more of explanation.